### PR TITLE
Added optional argument on which address to bind (socket).

### PR DESCRIPTION
**Pull request by binary1248**
Supersedes #678.

Should contain everything that was discussed in the previous pull request.

### CODE REVIEW

File: src/SFML/Network/IpAddress.cpp [hunk 1]
```diff
@@ -74,13 +79,15 @@ namespace sf
...
 
 
 ////////////////////////////////////////////////////////////
 IpAddress::IpAddress() :
-m_address(0)
+m_address(0),
+m_valid  (false)
 {
     // We're using 0 (INADDR_ANY) instead of INADDR_NONE to represent the invalid address,
     // because the latter is also the broadcast address (255.255.255.255); it's ok because
```
Comment (LaurentGomila) [inline src/SFML/Network/IpAddress.cpp]: Can't this comment be removed now?

**Comment by LaurentGomila** [comment]
The ```resolve``` function was useful when it returned a single integer that could be assigned in the initializer list of the constructors, but now I think that a private member function would be more readable:
* no need to return anything, you can directly assign members
* slightly less code duplication in constructors

The < operator uses ```m_valid``` and ```toInteger()```, since it is friend it could use ```m_adress```directly for consistency.

Implementing one of the comparison operators using the other would be slightly better than duplicating the logic. Personally I always do that for every comparable class that I write.

File: src/SFML/Network/IpAddress.cpp (hunk 1 above)
Comment (binary1248) [inline src/SFML/Network/IpAddress.cpp]: Hmm... I guess it can. :stuck_out_tongue:

**Comment by binary1248** [comment]
Pushed new version. Anything else?

**Comment by LaurentGomila** [comment]
Shouldn't ```resolve``` reset the members if it fails, instead of implicitly keeping their previous values? This would prevent potential mistakes if in the future we call this function in another context.

**Comment by binary1248** [comment]
Done. Anything else?

**Comment by LaurentGomila** [comment]
Just nitpicking: the return statements in ```resolve``` are not needed anymore. The 3rd one too, if the last piece of code of the function is put in a else block after testing the inet_addr result.

Otherwise, :+1: for me.

**Comment by binary1248** [comment]
Done.

**Comment by eXpl0it3r** [comment]
Looks fine to me, could need some further testing!

File: src/SFML/Network/IpAddress.cpp [hunk 2]
```diff
@@ -194,23 +160,72 @@ IpAddress IpAddress::getPublicAddress(Time timeout)
...
-    return left.toInteger() == right.toInteger();
+    return !(left < right) && !(right < left);
 }
 
 
 ////////////////////////////////////////////////////////////
 bool operator !=(const IpAddress& left, const IpAddress& right)
 {
-    return !(left == right);
+    return (left < right) || (right < left);
```
Comment (Bromeon) [inline src/SFML/Network/IpAddress.cpp]: Why not keep `return !(left == right);`? This is something that must always be true, and it's the simplest way of ensuring correct behavior, given `operator==` is implemented correctly. If the equality semantics change in the future, you don't need to touch both operands.

I assume you did that for performance reasons, but to be consequent, you'd rather implement `operator==` directly and not via `operator<` :wink:

File: src/SFML/Network/IpAddress.cpp [hunk 3]
```diff
@@ -194,23 +160,72 @@ IpAddress IpAddress::getPublicAddress(Time timeout)
...
 }
 
 
 ////////////////////////////////////////////////////////////
 bool operator <(const IpAddress& left, const IpAddress& right)
 {
-    return left.toInteger() < right.toInteger();
+    return (left.m_valid < right.m_valid) || (left.m_address < right.m_address);
```
Comment (Bromeon) [inline src/SFML/Network/IpAddress.cpp]: This violates [strict weak ordering](https://en.wikipedia.org/wiki/Weak_ordering#Strict_weak_orderings) and makes `sf::IpAddress` unusable in associative containers.

For example, consider the following objects:
```cpp
IpAddress a { m_valid = true, m_address = 0 };
IpAddress b { m_valid = false, m_address = 1 };
```
Both `a < b` and `b < a` are true.

You should always be overly cautious when using logical operators to implement relational operators. It's so easy to break behavior. A fool-proof and simple approach is to delegate the comparison to the standard library:
```cpp
// C++98
return std::make_pair(left.m_valid, left.m_address) < std::make_pair(right.m_valid, right.m_address);

// C++11
return std::make_tuple(/* all left attributes */) < std::make_pair(/* all right attributes */); // by value
return std::tie(/* all left attributes */) < std::tie(/* all right attributes */); // by reference
```

File: src/SFML/Network/IpAddress.cpp [hunk 4]
```diff
@@ -194,23 +160,72 @@ IpAddress IpAddress::getPublicAddress(Time timeout)
...
+    }
+    else if (address == "0.0.0.0")
+    {
+        m_address = INADDR_ANY;
+        m_valid = true;
+    }
+    else
+    {
+        // Try to convert the address as a byte representation ("xxx.xxx.xxx.xxx")
+        sf::Uint32 ip = inet_addr(address.c_str());
```
Comment (Bromeon) [inline src/SFML/Network/IpAddress.cpp]: Unnecessary `sf::` prefix

**Comment by binary1248** [comment]
Rebased and fixed.

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

File: src/SFML/Network/IpAddress.cpp [hunk 5]
```diff
@@ -210,7 +225,16 @@ bool operator !=(const IpAddress& left, const IpAddress& right)
...
+        return true;
+
+    if (right.m_valid < left.m_valid)
+        return false;
+
+    if (left.m_address < right.m_address)
+        return true;
+
+    return false;
 }
```
Comment (Bromeon) [inline src/SFML/Network/IpAddress.cpp]: Just wondering, is there a reason why you prefer this complicated case differentiation over a simple
```cpp
return std::make_pair(left.m_valid, left.m_address)
     < std::make_pair(right.m_valid, right.m_address);
```
as suggested? The latter is much more concise, clear and easier to maintain/extend.

Comment (LaurentGomila) [inline src/SFML/Network/IpAddress.cpp]: > easier to maintain/extend

How to you easily extend it if you add a 3rd member, with C++03 code?

Comment (binary1248) [inline src/SFML/Network/IpAddress.cpp]: The operations are equivalent.

Definition of `std::pair<T1, T2>::operator<` from the standard:
> Returns: `x.first < y.first || (!(y.first < x.first) && x.second < y.second)`.

Comment (Bromeon) [inline src/SFML/Network/IpAddress.cpp]: > How to you easily extend it if you add a 3rd member, with C++03 code?

```cpp
return std::make_pair(std::make_pair(left.m_valid, left.m_address), left.m_third)
     < std::make_pair(std::make_pair(right.m_valid, right.m_address), right.m_third);
```
Now, how do you easily extend binary1248's code to a third member? :grimacing: 

> The operations are equivalent.

I know, but shouldn't we go for the shorter and easier-to-understand code if we have two options of writing it down? Or what's the deal with `std::pair`? :wink:

File: include/SFML/Network/IpAddress.hpp [hunk 6]
```diff
@@ -182,15 +182,27 @@ class SFML_NETWORK_API IpAddress
     // Static member data
     ////////////////////////////////////////////////////////////
     static const IpAddress None;      ///< Value representing an empty/invalid address
+    static const IpAddress Any;       ///< Value representing any address (0.0.0.0)
     static const IpAddress LocalHost; ///< The "localhost" address (for connecting a computer to itself locally)
     static const IpAddress Broadcast; ///< The "broadcast" address (for sending UDP messages to everyone on a local network)
 
 private:
 
+    friend bool operator <(const IpAddress& left, const IpAddress& right);
```
Comment (Bromeon) [inline include/SFML/Network/IpAddress.hpp]: The friend declaration must have the same signature **and linkage specification** as the operator. This includes `SFML_NETWORK_API`.

That's also why the build currently fails on Visual Studio 2015.

Comment (binary1248) [inline include/SFML/Network/IpAddress.hpp]: Fixed.

File: src/SFML/Network/IpAddress.cpp (hunk 5 above)
Comment (binary1248) [inline src/SFML/Network/IpAddress.cpp]: So... roll our own or `std::make_pair`? There isn't that big of a difference unless you are using a really crappy compiler that doesn't know how to optimize out the pair constructions. I don't mind either of them and just chose the one with the guaranteed performance advantage.

@LaurentGomila, @Bromeon

Comment (LaurentGomila) [inline src/SFML/Network/IpAddress.cpp]: I don't mind, both are ok for me too.

Comment (binary1248) [inline src/SFML/Network/IpAddress.cpp]: Fixed.

**Comment by binary1248** [comment]
Anything else?

File: src/SFML/Network/IpAddress.cpp (hunk 5 above)
Comment (Bromeon) [inline src/SFML/Network/IpAddress.cpp]: I usually prefer simple and readable code over potential micro-optimizations. And when talking about performance, you should first optimize `operator==` to not call `operator<` twice :wink:
(don't, I'm just saying)

**Pull request merged**